internal/factsengine/factsengine.go

```diff
@@ -0,0 +1,174 @@
+package factsengine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/internal/factsengine/adapters"
+	"github.com/trento-project/agent/internal/factsengine/gatherers"
+)
+
+type factsEngine struct {
+	agentID             string
+	factsEngineService  string
+	factGatherers       map[string]gatherers.FactGatherer
+	factsServiceAdapter adapters.Adapter
+}
+
+func NewFactsEngine(agentID, factsEngineService string) *factsEngine {
+	return &factsEngine{
+		agentID:            agentID,
+		factsEngineService: factsEngineService,
+		factGatherers: map[string]gatherers.FactGatherer{
+			gatherers.CorosyncFactKey: gatherers.NewCorosyncConfGatherer(),
+		},
+	}
+}
+
+func (c *factsEngine) Subscribe() error {
+	log.Infof("Subscribing agent %s to the facts gathering reception service on %s", c.agentID, c.factsEngineService)
+	//RabbitMQ adapter exists only by now
+	c.factsServiceAdapter = adapters.NewRabbitMQAdapter(c.factsEngineService)
+	log.Infof("Subscription to the facts engine by agent %s in %s done", c.agentID, c.factsEngineService)
+
+	return nil
+}
+
+func (c *factsEngine) Unsubscribe() error {
+	log.Infof("Unsubscribing agent %s from the facts engine service", c.agentID)
+	c.factsServiceAdapter.Unsubscribe()
+	log.Infof("Unsubscribed properly")
+
+	return nil
+}
+
+func (c *factsEngine) Listen(ctx context.Context) {
+	log.Infof("Listening for facts gathering events...")
+	defer c.Unsubscribe()
+
+	err := c.factsServiceAdapter.Listen(c.agentID, c.handleRequest)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *factsEngine) handleRequest(request []byte) error {
+	factsRequests, err := parseFactsRequest(request)
+	if err != nil {
+		log.Errorf("Invalid facts request: %s", err)
+		return err
+	}
+
+	gatheredFacts, err := gatherFacts(factsRequests, c.factGatherers)
+	if err != nil {
+		log.Errorf("Error gathering facts: %s", err)
+		return err
+	}
+
+	if err := c.publishFacts(gatheredFacts); err != nil {
+		log.Errorf("Error publishing facts: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+func gatherFacts(groupedFactsRequest *gatherers.GroupedFactsRequest, factGatherers map[string]gatherers.FactGatherer) (*gatherers.FactsResult, error) {
```

Comment:
maybe we can use a errgroup here to coordinate all the goroutines and do one error handling and one waiting